src/or/circuitbuild.c

```diff
@@ -2464,7 +2464,7 @@ count_acceptable_nodes, (smartlist_t *nodes))
     if (! node->is_valid)
 //      log_debug(LD_CIRC,"Nope, the directory says %d is not valid.",i);
       continue;
-    if (! node_has_descriptor(node))
+    if (! node_has_any_descriptor(node))
```

Comment:
I think this code is correct, but it's not obvious:
1. new_route_len() is the only caller
2. new_route_len() will fail if count_acceptable_nodes() is less than the route length.
3. later functions will fail if there is no guard
4. later functions will fail if there are not enough subsequent nodes for the rest of the route

So this check can spuriously succeed when we have no guards, or when we have many bridges, and no subsequent nodes. But it can never fail when there are actually enough nodes for the path. So this is just an optimisation to stop us trying to build lots of circuits when we have no descriptors.

Reply:
We could return the number that pass for_direct_connect = 1 and for_direct_connect = 0, but that seems unnecessarily complicated.

Reply:
I've opened trac:25885 to change this if we want to do so someday.   I've added a comment to note this defect as df7eaf84439363.